packages/lodestar/src/chain/chain.ts

```diff
@@ -193,6 +202,92 @@ export class BeaconChain extends (EventEmitter as { new(): ChainEventEmitter })
+    let firstSLot = firstBlock.message.slot;
+    let lastSlot = lastBlock.message.slot;
+    this.logger.info(`Found ${allBlocks.length} nonfinalized blocks in database from slot ${firstSLot} to ${lastSlot}`);
+    // initially we initialize database with genesis block
+    if (allBlocks.length === 1) {
+      // start from scratch
+      const blockHash = this.config.types.BeaconBlock.hashTreeRoot(firstBlock.message);
+      if (this.config.types.Root.equals(blockHash, this.forkChoice.headBlockRoot())) {
+        this.logger.info("Chain is up to date, no need to restore");
+        return;
+      }
+    }
+    const isCheckpointNotGenesis = lastKnownState.slot > GENESIS_SLOT;
+    const stateRoot = this.config.types.BeaconState.hashTreeRoot(lastKnownState);
+    const finalizedBlock = sortedBlocks.find(block => {
+      return (block.message.slot === lastKnownState.slot) &&
+      // at genesis the genesis block's state root is not equal to genesis state root
+      (isCheckpointNotGenesis? this.config.types.Root.equals(block.message.stateRoot, stateRoot) : true);
+    });
+    if (!finalizedBlock) {
+      throw new Error(`Cannot find block for finalized state at slot ${lastKnownState.slot}`);
+    } else {
+      this.logger.info(`Found finalized block at slot ${finalizedBlock.message.slot},
+        root=${toHexString(this.config.types.BeaconBlock.hashTreeRoot(finalizedBlock.message))}`);
+    }
+    // init forkchoice
+    const blockCheckpoint: Checkpoint = {
+      root: this.config.types.BeaconBlock.hashTreeRoot(finalizedBlock.message),
+      epoch: computeEpochAtSlot(this.config, finalizedBlock.message.slot)
+    };
+    this.forkChoice.addBlock({
+      slot: finalizedBlock.message.slot,
+      blockRoot: this.config.types.BeaconBlock.hashTreeRoot(finalizedBlock.message),
+      stateRoot: finalizedBlock.message.stateRoot.valueOf() as Uint8Array,
+      parentRoot: finalizedBlock.message.parentRoot.valueOf() as Uint8Array,
+      justifiedCheckpoint: (isCheckpointNotGenesis)? lastKnownState.currentJustifiedCheckpoint : blockCheckpoint,
+      finalizedCheckpoint: (isCheckpointNotGenesis)? lastKnownState.finalizedCheckpoint : blockCheckpoint
+    });
+    // init state cache
+    this.db.stateCache.add(lastKnownState);
+    // no need to process the finalized block
+    const processedBlocks = sortedBlocks.filter((block) => block.message.slot > finalizedBlock.message.slot);
+    if (!processedBlocks.length) {
+      this.logger.info("No need to process blocks");
+      return;
+    }
+    firstSLot = processedBlocks[0].message.slot;
+    lastSlot = processedBlocks[processedBlocks.length - 1].message.slot;
+    this.logger.info(`Start processing from slot ${firstSLot} to ${lastSlot} to rebuild state cache and forkchoice`);
+    for (const block of processedBlocks) {
+      await this.receiveBlock(block, true, true);
+    }
+    await this.waitForBlockProcessed(this.config.types.BeaconBlock.hashTreeRoot(lastBlock.message));
```

Comment:
might be a race condition here?
where the block processor processes the lastBlock before the processedBlock event is subscribed to?